File: runner/models/model_codegen.py
import openai
import os
import textwrap
import time
import random
from . import ConvexCodegenModel, SYSTEM_PROMPT, ModelTemplate, ModelProvider
from markdown_it import MarkdownIt
from typing import Union
from .guidelines import Guideline, GuidelineSection, CONVEX_GUIDELINES
from braintrust import wrap_openai


# Retry configuration for flaky API providers (e.g., Together.xyz 503 errors)
MAX_RETRIES = 5
INITIAL_RETRY_DELAY_SECONDS = 2.0
MAX_RETRY_DELAY_SECONDS = 60.0
RETRY_JITTER_FACTOR = 0.25  # Add up to 25% random jitter


def should_skip_guidelines() -> bool:
    """Check if guidelines should be skipped based on EVALS_EXPERIMENT env var."""
    return os.getenv("EVALS_EXPERIMENT") == "no_guidelines"


def get_guidelines_content() -> str:
    """Get guidelines content from custom file or default."""
    custom_path = os.getenv("CUSTOM_GUIDELINES_PATH")
    if custom_path and os.path.exists(custom_path):
        with open(custom_path, "r") as f:
            return f.read()
    if should_skip_guidelines():
        return ""
    return "".join(render_guidelines(CONVEX_GUIDELINES))


class Model(ConvexCodegenModel):
    def __init__(self, api_key: str, model: ModelTemplate):
        self.model = model
        # Allow disabling Braintrust proxy entirely via env toggle
        disable_proxy = os.getenv("DISABLE_BRAINTRUST") == "1"

        url = "https://api.braintrust.dev/v1/proxy"

        if disable_proxy:
            match model.provider:
                case ModelProvider.OPENAI:
                    url = "https://api.openai.com/v1"
                case ModelProvider.ANTHROPIC:
                    url = "https://api.anthropic.com/v1"
                case ModelProvider.TOGETHER:
                    url = "https://api.together.xyz/v1"
                case ModelProvider.GOOGLE:
                    url = "https://generativelanguage.googleapis.com/v1beta"
                case ModelProvider.XAI:
                    url = "https://api.x.ai/v1"
                case ModelProvider.MOONSHOT:
                    url = "https://api.moonshot.ai/v1"
                case _:
                    raise ValueError(f"Unknown model provider for disable-proxy mode: {model.provider}")

        # Configure OpenAI client with increased retries for transient errors
        if model.override_proxy:
            url = model.override_proxy
            client = openai.OpenAI(
                base_url=url,
                api_key=api_key,
                max_retries=MAX_RETRIES,
                timeout=openai.Timeout(connect=30.0, read=300.0, write=30.0, pool=30.0),
            )
        else:
            base_client = openai.OpenAI(
                base_url=url,
                api_key=api_key,
                max_retries=MAX_RETRIES,
                timeout=openai.Timeout(connect=30.0, read=300.0, write=30.0, pool=30.0),
            )
            client = base_client if disable_proxy else wrap_openai(base_client)
        self.client = client

    def generate(self, prompt: str):
        user_prompt = "".join(render_prompt(self.model.requires_chain_of_thought, prompt))
        
        # Use Responses API for models that require it (e.g., gpt-5.2-codex)
        if self.model.uses_responses_api:
            return self._generate_with_responses_api(user_prompt)
        
        if self.model.uses_system_prompt:
            system_message = {"role": "system", "content": SYSTEM_PROMPT}
        else:
            system_message = {"role": "user", "content": SYSTEM_PROMPT}
        # Build parameters, selecting the correct token limit key for newer models
        # Together AI models (DeepSeek, Llama) have a 12289 total context limit,
        # so we use a lower max_tokens to leave room for input tokens (~7000)
        if self.model.provider == ModelProvider.TOGETHER:
            max_token_limit = 4096
        elif self.model.name == "claude-3-5-sonnet-latest":
            max_token_limit = 8192
        else:
            max_token_limit = 16384
        create_params = {
            "model": self.model.name,
            "messages": [
                system_message,
                {"role": "user", "content": user_prompt},
            ],
        }
        # Only add temperature for models that support it (reasoning models like o1/o3/gpt-5 don't)
        if self.model.supports_temperature:
            temperature = float(os.getenv("EVAL_TEMPERATURE", "0.7"))
            create_params["temperature"] = temperature
        # Some newer models (e.g., GPT-5 family, o4) expect `max_completion_tokens` instead of `max_tokens`.
        if self.model.name.startswith("gpt-5") or self.model.name.startswith("o4"):
            create_params["max_completion_tokens"] = max_token_limit
        else:
            create_params["max_tokens"] = max_token_limit

        response = self._call_with_retry(create_params)
        return self._parse_response(response.choices[0].message.content)
    
    def _generate_with_responses_api(self, user_prompt: str):
        """
        Generate using the OpenAI Responses API for models like gpt-5.2-codex.
        
        The Responses API uses a different endpoint (/v1/responses) and request format:
        - 'instructions' for system-level guidance
        - 'input' for user content (can be a string or list of messages)
        - Response text is in response.output_text
        """
        max_token_limit = 16384
        
        create_params = {
            "model": self.model.name,
            "instructions": SYSTEM_PROMPT,
            "input": user_prompt,
            "max_output_tokens": max_token_limit,
            "store": False,  # Don't persist the response
        }
        
        response = self._call_responses_api_with_retry(create_params)
        return self._parse_response(response.output_text)

    def _call_with_retry(self, create_params: dict):
        """
        Call the API with additional retry logic for transient errors.
        
        The OpenAI client has built-in retries, but for very flaky providers
        (like Together.xyz which often returns 503), we add an outer retry loop
        with exponential backoff.
        """
        last_exception = None
        delay = INITIAL_RETRY_DELAY_SECONDS

        for attempt in range(MAX_RETRIES):
            try:
                return self.client.chat.completions.create(**create_params)
            except openai.APIStatusError as e:
                # Retry on 5xx server errors and 429 rate limits
                if e.status_code in (429, 500, 502, 503, 504):
                    last_exception = e
                    # Add jitter to prevent thundering herd
                    jitter = delay * RETRY_JITTER_FACTOR * random.random()
                    sleep_time = min(delay + jitter, MAX_RETRY_DELAY_SECONDS)
                    print(f"API error {e.status_code}, retrying in {sleep_time:.1f}s (attempt {attempt + 1}/{MAX_RETRIES})")
                    time.sleep(sleep_time)
                    delay *= 2  # Exponential backoff
                else:
                    raise
            except openai.APIConnectionError as e:
                # Retry on connection errors
                last_exception = e
                jitter = delay * RETRY_JITTER_FACTOR * random.random()
                sleep_time = min(delay + jitter, MAX_RETRY_DELAY_SECONDS)
                print(f"Connection error, retrying in {sleep_time:.1f}s (attempt {attempt + 1}/{MAX_RETRIES})")
                time.sleep(sleep_time)
                delay *= 2

        # If we've exhausted all retries, raise the last exception
        raise last_exception

    def _call_responses_api_with_retry(self, create_params: dict):
        """
        Call the OpenAI Responses API with retry logic for transient errors.
        
        Similar to _call_with_retry but uses the responses.create endpoint
        instead of chat.completions.create.
        """
        last_exception = None
        delay = INITIAL_RETRY_DELAY_SECONDS

        for attempt in range(MAX_RETRIES):
            try:
                return self.client.responses.create(**create_params)
            except openai.APIStatusError as e:
                # Retry on 5xx server errors and 429 rate limits
                if e.status_code in (429, 500, 502, 503, 504):
                    last_exception = e
                    jitter = delay * RETRY_JITTER_FACTOR * random.random()
                    sleep_time = min(delay + jitter, MAX_RETRY_DELAY_SECONDS)
                    print(f"Responses API error {e.status_code}, retrying in {sleep_time:.1f}s (attempt {attempt + 1}/{MAX_RETRIES})")
                    time.sleep(sleep_time)
                    delay *= 2
                else:
                    raise
            except openai.APIConnectionError as e:
                last_exception = e
                jitter = delay * RETRY_JITTER_FACTOR * random.random()
                sleep_time = min(delay + jitter, MAX_RETRY_DELAY_SECONDS)
                print(f"Connection error, retrying in {sleep_time:.1f}s (attempt {attempt + 1}/{MAX_RETRIES})")
                time.sleep(sleep_time)
                delay *= 2

        raise last_exception

    def _parse_response(self, response: str):
        md = MarkdownIt()
        tokens = md.parse(response)

        files = {}
        current_file = None
        in_files_section = False

        for i, token in enumerate(tokens):
            if token.type == "heading_open" and token.tag == "h1":
                title_token = tokens[i + 1]
                if title_token.content == "Files":
                    in_files_section = True
                    continue

            if not in_files_section:
                continue

            if token.type == "heading_open" and token.tag == "h2":
                title_token = tokens[i + 1]
                current_file = title_token.content.strip()
            elif token.type == "fence" and current_file:
                files[current_file] = token.content.strip()
                current_file = None

        return files


def render_prompt(chain_of_thought: bool, task_description: str):
    yield "Your task is to generate a Convex backend from a task description.\n"
    if chain_of_thought:
        yield "Before writing any code, analyze the task and think through your approach. Use the Analysis section to show your thought process, covering the following areas:\n"
        yield "1. Summarize the task requirements\n"
        yield "2. List out the main components needed for the backend\n"
        yield "3. Design the public API and internal functions:\n"
        yield "   - List each function with its file path, argument validators, and return validator, and purpose.\n"
        yield "4. Plan the schema design (if needed):\n"
        yield "   - List each table with its validator (excluding the included _id and _creationTime fields) and its indexes\n"
        yield "5. Outline background processing requirements (if any):\n"
        yield "After your analysis, output all files within an h1 Files section that has an h2 section for each necessary file for a Convex backend that implements the requested functionality.\n"
        yield "For example, correct output looks like\n"
        yield "# Analysis\n"
        yield "...\n"
        yield "# Files\n"
        yield "## package.json\n"
        yield "```\n"
        yield "...\n"
        yield "```\n"
        yield "## tsconfig.json\n"
        yield "```\n"
        yield "...\n"
        yield "```\n"
        yield "## convex/schema.ts\n"
        yield "```\n"
        yield "...\n"
        yield "```\n"
    else:
        yield "Output all files within an h1 Files section that has an h2 section for each necessary file for a Convex backend that implements the requested functionality.\n"
        yield "For example, correct output looks like\n"
        yield "# Files\n"
        yield "## package.json\n"
        yield "```\n"
        yield "...\n"
        yield "```\n"
        yield "## tsconfig.json\n"
        yield "```\n"
        yield "...\n"
        yield "```\n"
        yield "## convex/schema.ts\n"
        yield "```\n"
        yield "...\n"
        yield "```\n"

    yield from render_examples()
    yield "\n"

    yield "# General Coding Standards\n"
    yield "- Use 2 spaces for code indentation.\n"
    yield "- Ensure your code is clear, efficient, concise, and innovative.\n"
    yield "- Maintain a friendly and approachable tone in any comments or documentation.\n\n"

    guidelines_content = get_guidelines_content()
    if guidelines_content:
        yield guidelines_content
        yield "\n"

    yield "\n# File Structure\n"
    yield "- You can write to `package.json`, `tsconfig.json`, and any files within the `convex/` folder.\n"
    yield "- Do NOT write to the `convex/_generated` folder. You can assume that `npx convex dev` will populate this folder.\n"
    yield "- It's VERY IMPORTANT to output files to the correct paths, as specified in the task description.\n"
    yield "- Always start with `package.json` and `tsconfig.json` files.\n"
    yield '- Use Convex version "^1.31.2".\n\n'
    yield '- Use Typescript version "^5.7.3".\n\n'

    if chain_of_thought:
        yield "Begin your response with your thought process, then proceed to generate the necessary files for the Convex backend.\n"

    yield "Now, implement a Convex backend that satisfies the following task description:\n"
    yield f"```\n{task_description}\n```\n"


def render_examples():
    yield "# Examples:\n"
    for example in os.listdir("examples/"):
        example_path = os.path.join("examples/", example)
        if not os.path.isdir(example_path):
            continue

        task_description = open(os.path.join(example_path, "TASK.txt"), "r").read()
        analysis = open(os.path.join(example_path, "ANALYSIS.txt"), "r").read()

        file_paths = []
        for dirpath, _, file_names in os.walk(example_path, topdown=True):
            if "node_modules" in dirpath or "_generated" in dirpath:
                continue
            for file_name in file_names:
                if (
                    file_name == "package.json"
                    or file_name == "tsconfig.json"
                    or file_name.endswith(".ts")
                    or file_name.endswith(".tsx")
                ):
                    file_paths.append(os.path.join(dirpath, file_name))

        file_paths.sort(key=lambda x: (x.count("/"), x))

        yield f"## Example: {example}\n\n"
        yield "### Task\n"
        yield f"```\n{task_description}\n```\n\n"
        yield "### Analysis\n"
        yield f"{analysis}\n\n"
        yield "### Implementation\n\n"
        for file_path in file_paths:
            rel_path = os.path.relpath(file_path, example_path)
            file_content = open(file_path, "r").read()
            file_content = textwrap.dedent(file_content).strip()
            yield f"#### {rel_path}\n"
            yield f"```typescript\n{file_content}\n```\n\n"


def render_guidelines(node: Union[GuidelineSection, Guideline], header="#"):
    if isinstance(node, Guideline):
        yield "- "
        yield node.content
        yield "\n"
    else:
        words = node.name.split("_")
        words[0] = words[0].capitalize()
        yield f"{header} {' '.join(words)}\n"
        for child in node.children:
            yield from render_guidelines(child, header + "#")
        yield "\n"


# Used by the eval system
OPENAI_CONVEX_GUIDELINES = "".join(render_guidelines(CONVEX_GUIDELINES))


def build_release_rules() -> str:
    return "".join(render_guidelines(CONVEX_GUIDELINES)) + "".join(render_examples())

File: runner/models/modelCodegen.ts
/**
 * LLM code generation: builds prompts, calls provider APIs, parses responses.
 *
 * Uses the Vercel AI SDK (generateText) as a unified interface across all
 * providers. When the "web_search" experiment is active, a Tavily-powered
 * search tool is made available to every model.
 */
import { generateText, type LanguageModel } from "ai";
import { createOpenAI } from "@ai-sdk/openai";
import { createOpenAICompatible } from "@ai-sdk/openai-compatible";
import MarkdownIt from "markdown-it";
import { readFileSync, readdirSync, statSync, existsSync } from "fs";
import { join, relative } from "path";
import {
  type ModelTemplate,
  OPENROUTER_BASE_URL,
  SYSTEM_PROMPT,
} from "./index.js";
import {
  CONVEX_GUIDELINES,
  renderCompactGuidelines,
  renderGuidelines,
} from "./guidelines.js";
import {
  webSearchTool,
  WEB_SEARCH_SYSTEM_SUPPLEMENT,
  MAX_TOOL_STEPS,
} from "./webSearch.js";
import { logInfo } from "../logging.js";
import { stepCountIs } from "ai";

// ── Experiment helpers ────────────────────────────────────────────────

function isWebSearchEnabled(): boolean {
  const exp = process.env.EVALS_EXPERIMENT;
  return exp === "web_search" || exp === "web_search_no_guidelines";
}

// ── Guidelines helpers ────────────────────────────────────────────────

function getGuidelinesContent(): string {
  const customPath = process.env.CUSTOM_GUIDELINES_PATH;
  if (customPath && existsSync(customPath)) {
    return readFileSync(customPath, "utf-8");
  }
  const exp = process.env.EVALS_EXPERIMENT;
  if (exp === "no_guidelines" || exp === "web_search_no_guidelines") return "";
  if (exp === "agents_md") return renderCompactGuidelines();
  return renderGuidelines(CONVEX_GUIDELINES);
}

// ── AI SDK model construction ────────────────────────────────────────

/**
 * Create an AI SDK LanguageModel from our ModelTemplate + API key.
 * All models are accessed via OpenRouter.
 */
function createLanguageModel(
  template: ModelTemplate,
  apiKey: string,
): LanguageModel {
  const baseURL = template.overrideProxy ?? OPENROUTER_BASE_URL;

  // Models using the Responses API need the OpenAI SDK pointed at
  // OpenRouter's /responses endpoint.
  if (template.apiKind === "responses") {
    const openai = createOpenAI({ apiKey, baseURL });
    return openai.responses(template.name);
  }

  const openrouter = createOpenAICompatible({
    name: "openrouter",
    baseURL,
    apiKey,
  });
  return openrouter.chatModel(template.name);
}

// ── Token limit helpers ──────────────────────────────────────────────

function getMaxOutputTokens(template: ModelTemplate): number {
  // Former Together models (DeepSeek) had 4096 limit
  if (template.name.startsWith("deepseek/")) return 4096;
  if (template.name === "anthropic/claude-3.5-sonnet") return 8192;
  return 16384;
}

// ── Model class ───────────────────────────────────────────────────────

export class Model {
  private languageModel: LanguageModel;
  private template: ModelTemplate;

  constructor(apiKey: string, model: ModelTemplate) {
    this.template = model;
    this.languageModel = createLanguageModel(model, apiKey);
  }

  async generate(prompt: string): Promise<Record<string, string>> {
    const userPrompt = renderPrompt(prompt);
    const useWebSearch = isWebSearchEnabled();

    const systemContent = useWebSearch
      ? `${SYSTEM_PROMPT}\n\n${WEB_SEARCH_SYSTEM_SUPPLEMENT}`
      : SYSTEM_PROMPT;

    const maxTokens = getMaxOutputTokens(this.template);

    // Build the base options shared across both prompt styles
    const baseOptions = {
      model: this.languageModel,
      maxOutputTokens: maxTokens,
      maxRetries: 5,
    };

    const promptOptions = {
      system: systemContent,
      prompt: userPrompt,
    };

    const options: Parameters<typeof generateText>[0] = {
      ...baseOptions,
      ...promptOptions,
    };

    // When the web search experiment is active, provide the tool and
    // allow multiple steps so the model can search then generate.
    if (useWebSearch) {
      options.tools = { web_search: webSearchTool };
      options.stopWhen = stepCountIs(MAX_TOOL_STEPS);
      options.onStepFinish = ({ toolCalls }) => {
        if (toolCalls && toolCalls.length > 0) {
          logInfo(
            `  [web_search] Model made ${toolCalls.length} tool call(s)`,
          );
        }
      };
    }

    const result = await generateText(options);
    return parseMarkdownResponse(result.text);
  }
}

// ── Response parsing ──────────────────────────────────────────────────

export function parseMarkdownResponse(
  response: string,
): Record<string, string> {
  const md = new MarkdownIt();
  const tokens = md.parse(response, {});

  const files: Record<string, string> = {};
  let currentFile: string | null = null;
  let inFilesSection = false;

  for (let i = 0; i < tokens.length; i++) {
    const token = tokens[i];

    if (token.type === "heading_open" && token.tag === "h1") {
      const titleToken = tokens[i + 1];
      if (titleToken?.content === "Files") {
        inFilesSection = true;
        continue;
      }
    }

    if (!inFilesSection) continue;

    if (token.type === "heading_open" && token.tag === "h2") {
      const titleToken = tokens[i + 1];
      currentFile = titleToken?.content?.trim() ?? null;
    } else if (token.type === "fence" && currentFile) {
      files[currentFile] = token.content.trim();
      currentFile = null;
    }
  }

  return files;
}

// ── Prompt rendering ──────────────────────────────────────────────────

const FILE_FORMAT_EXAMPLE = [
  "# Files",
  "## package.json",
  "```\n...\n```",
  "## tsconfig.json",
  "```\n...\n```",
  "## convex/schema.ts",
  "```\n...\n```",
].join("\n");

export function renderPrompt(taskDescription: string): string {
  const sections: string[] = [
    "Your task is to generate a Convex backend from a task description.",
  ];

  sections.push(
    `Output all files within an h1 Files section that has an h2 section for each necessary file for a Convex backend that implements the requested functionality.
For example, correct output looks like
${FILE_FORMAT_EXAMPLE}`,
  );

  sections.push(renderExamples());

  sections.push(`# General Coding Standards
- Use 2 spaces for code indentation.
- Ensure your code is clear, efficient, concise, and innovative.
- Maintain a friendly and approachable tone in any comments or documentation.`);

  const guidelinesContent = getGuidelinesContent();
  if (guidelinesContent) {
    sections.push(guidelinesContent);
  }

  sections.push(`# File Structure
- You can write to \`package.json\`, \`tsconfig.json\`, and any files within the \`convex/\` folder.
- Do NOT write to the \`convex/_generated\` folder. You can assume that \`npx convex dev\` will populate this folder.
- It's VERY IMPORTANT to output files to the correct paths, as specified in the task description.
- Always start with \`package.json\` and \`tsconfig.json\` files.
- Use Convex version "^1.31.2".
- Use Typescript version "^5.7.3".`);

  sections.push(
    `Now, implement a Convex backend that satisfies the following task description:\n\`\`\`\n${taskDescription}\n\`\`\``,
  );

  return sections.join("\n\n") + "\n";
}

function renderExamples(): string {
  const examplesDir = "examples";
  if (!existsSync(examplesDir)) return "";

  const parts: string[] = ["# Examples:"];

  for (const example of readdirSync(examplesDir)) {
    const examplePath = join(examplesDir, example);
    if (!statSync(examplePath).isDirectory()) continue;

    const taskDescription = readFileSync(join(examplePath, "TASK.txt"), "utf-8");
    const analysis = readFileSync(join(examplePath, "ANALYSIS.txt"), "utf-8");
    const filePaths = collectExampleFiles(examplePath);

    parts.push(`## Example: ${example}\n`);
    parts.push(`### Task\n\`\`\`\n${taskDescription}\n\`\`\`\n`);
    parts.push(`### Analysis\n${analysis}\n`);
    parts.push("### Implementation\n");

    for (const filePath of filePaths) {
      const relPath = relative(examplePath, filePath).replace(/\\/g, "/");
      const content = readFileSync(filePath, "utf-8").trim();
      parts.push(`#### ${relPath}\n\`\`\`typescript\n${content}\n\`\`\`\n`);
    }
  }

  return parts.join("\n");
}

/** Collect relevant source files from an example directory, sorted by depth then name. */
function collectExampleFiles(examplePath: string): string[] {
  const filePaths: string[] = [];
  walkDir(examplePath, (filePath) => {
    if (filePath.includes("node_modules") || filePath.includes("_generated")) {
      return;
    }
    const name = filePath.split(/[/\\]/).pop()!;
    if (
      name === "package.json" ||
      name === "tsconfig.json" ||
      name.endsWith(".ts") ||
      name.endsWith(".tsx")
    ) {
      filePaths.push(filePath);
    }
  });

  return filePaths.sort((a, b) => {
    const depthA = a.split(/[/\\]/).length;
    const depthB = b.split(/[/\\]/).length;
    return depthA !== depthB ? depthA - depthB : a.localeCompare(b);
  });
}

function walkDir(dir: string, callback: (path: string) => void): void {
  for (const entry of readdirSync(dir, { withFileTypes: true })) {
    const fullPath = join(dir, entry.name);
    if (entry.isDirectory()) {
      walkDir(fullPath, callback);
    } else {
      callback(fullPath);
    }
  }
}

/** Render guidelines + examples for release builds. */
export function buildReleaseRules(): string {
  return renderGuidelines(CONVEX_GUIDELINES) + renderExamples();
}
